torch/nested/__init__.py

```diff
@@ -1,11 +1,89 @@
+from typing import List, Optional
 import torch
 from torch._C import _add_docstr, _nested  # type: ignore[attr-defined]
+from torch import Tensor
+
+from torch.types import _dtype as DType
+from torch.types import _device as Device
 
 __all__ = [
-    'to_padded_tensor'
+    'to_padded_tensor',
+    'as_nested_tensor',
+    'nested_tensor',
 ]
 
-Tensor = torch.Tensor
+# Nested Tensor constructor functions
+# TODO: move these to pybind to accept numpy/nested lists as inputs in the future
+def nested_tensor(tensor_list: List[Tensor], *, dtype: Optional[DType] = None, device: Optional[Device] = None,
+                  requires_grad: Optional[bool] = False, pin_memory: Optional[bool] = False) -> Tensor:
+    r"""
+    Constructs a nested tensor with no autograd history (also known as a “leaf tensor”, see
+    :ref:`Autograd mechanics <autograd-mechanics>`) from :attr:`tensor_list` a list of tensors.
+
+    Args:
+        tensor_list (List[Tensor]): a list of tensors with the same ndim
+
+    Keyword arguments:
+        dtype (:class:`torch.dtype`, optional): the desired type of returned nested tensor.
+            Default: if None, same :class:`torch.dtype` as leftmost tensor in the list.
+        device (:class:`torch.device`, optional): the desired device of returned nested tensor.
+            Default: if None, same :class:`torch.device` as leftmost tensor in the list
+        requires_grad (bool, optional): If autograd should record operations on the
+            returned nested tensor. Default: ``False``.
+        pin_memory (bool, optional): If set, returned nested tensor would be allocated in
+            the pinned memory. Works only for CPU tensors. Default: ``False``.
+
+    Example::
+
+        >>> a = torch.arange(3, dtype=torch.float, requires_grad=True)
+        >>> b = torch.arange(5, dtype=torch.float, requires_grad=True)
+        >>> nt = torch.nested.nested_tensor([a, b], requires_grad=True)
+        >>> nt.is_leaf
+        True
+    """
+    if not isinstance(tensor_list, list) or any([not torch.is_tensor(t) for t in tensor_list]):
+        raise TypeError("nested_tensor(): Expected first argument to be a list of tensors ")
+    new_data = []
+    for t in tensor_list:
+        new_data.append(t.clone().detach())
```

Comment:
Small performance hit by having to clone but I think it is worth it in order to do this all in python

Reply:
Why do we have to clone?

Reply:
Should it just be `t.detach()`?

Reply:
Yeah I think that's enough

Reply:
Ah I see will change, I understand that `cat` in NestedTensorUtils:wrap_tensor_node will ultimately create a copy, but in practice I've never seen `detach` used without `clone`

Reply:
In the readme: 
```
# NOTE: Previously, if we change the tensor metadata (e.g. sizes / strides /
# storage / storage_offset) of a tensor created from `detach()`, those metadata
# in the original tensor will also be updated. However, the new behavior is that
# those metadata changes to the detached tensor will not update the original tensor
# anymore, and in the `detach()` function we need to set `allow_tensor_metadata_change_`
# to false to make such changes explicitly illegal, in order to prevent users from
# changing metadata of the detached tensor and expecting the original tensor to also
# be updated.
```
I think the concern is that we don't want the detach to mutate the original tensor but like you said since wrap_tensor makes a copy I think in this case it is okay.

Reply:
Ah gotcha thanks Driss :)